File: backend/internal/shadow/model_dictionary.go
package shadow

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"gopkg.in/yaml.v3"
)

// ModelField 模型字段。
type ModelField struct {
	Name     string `yaml:"name" json:"name"`
	Type     string `yaml:"type" json:"type"`
	Required bool   `yaml:"required" json:"required"`
}

// ModelRelationship 模型关系。
type ModelRelationship struct {
	Entity string `yaml:"entity" json:"entity"`
	DTO    string `yaml:"dto" json:"dto"`
	Type   string `yaml:"type" json:"type"` // map, extend, subset, transform
}

// ModelEntry 模型字典条目。
type ModelEntry struct {
	Name          string              `yaml:"name" json:"name"`
	Fields        []ModelField        `yaml:"fields" json:"fields"`
	Relationships []ModelRelationship `yaml:"relationships,omitempty" json:"relationships,omitempty"`
	Source        string              `yaml:"source" json:"source"`
	Tags          []string            `yaml:"tags" json:"tags"`
	Similarity    float64             `yaml:"similarity,omitempty" json:"similarity,omitempty"`
}

// ModelDictionaryConfig 模型字典配置。
type ModelDictionaryConfig struct {
	DictionaryPath      string
	SimilarityThreshold float64
}

// ModelDuplicateCheckResult 模型重复检查结果。
type ModelDuplicateCheckResult struct {
	IsDuplicate   bool
	SimilarModels []ModelEntry
}

// ModelDictionary 模型字典，防止重复数据结构。
type ModelDictionary struct {
	mu            sync.RWMutex
	models        []ModelEntry
	relationships []ModelRelationship
	config        ModelDictionaryConfig
}

// NewModelDictionary 创建模型字典实例。
func NewModelDictionary(config *ModelDictionaryConfig) *ModelDictionary {
	cfg := ModelDictionaryConfig{
		DictionaryPath:      ".codeflow/registry/models.yaml",
		SimilarityThreshold: 0.6,
	}
	if config != nil {
		if config.DictionaryPath != "" {
			cfg.DictionaryPath = config.DictionaryPath
		}
		if config.SimilarityThreshold > 0 {
			cfg.SimilarityThreshold = config.SimilarityThreshold
		}
	}
	return &ModelDictionary{
		models:        make([]ModelEntry, 0),
		relationships: make([]ModelRelationship, 0),
		config:        cfg,
	}
}

// Register 注册模型条目，检查重复后添加。
func (d *ModelDictionary) Register(entry ModelEntry) ModelDuplicateCheckResult {
	d.mu.Lock()
	defer d.mu.Unlock()

	result := d.checkDuplicateLocked(entry)
	if result.IsDuplicate {
		return result
	}

	entry.Similarity = 0
	d.models = append(d.models, entry)
	return ModelDuplicateCheckResult{IsDuplicate: false, SimilarModels: nil}
}

// Search 搜索匹配的模型。
func (d *ModelDictionary) Search(query string) []ModelEntry {
	d.mu.RLock()
	defer d.mu.RUnlock()

	queryKw := extractKeywords(query)
	if len(queryKw) == 0 {
		return nil
	}

	type scored struct {
		model ModelEntry
		score float64
	}

	var results []scored
	for _, model := range d.models {
		modelKw := modelKeywords(model)
		score := computeKeywordSimilarity(queryKw, modelKw)
		if score > 0 {
			m := model
			m.Similarity = score
			results = append(results, scored{model: m, score: score})
		}
	}

	for i := 0; i < len(results); i++ {
		for j := i + 1; j < len(results); j++ {
			if results[j].score > results[i].score {
				results[i], results[j] = results[j], results[i]
			}
		}
	}

	out := make([]ModelEntry, len(results))
	for i, r := range results {
		out[i] = r.model
	}
	return out
}

// CheckDuplicate 检查是否存在重复模型。
func (d *ModelDictionary) CheckDuplicate(entry ModelEntry) ModelDuplicateCheckResult {
	d.mu.RLock()
	defer d.mu.RUnlock()
	return d.checkDuplicateLocked(entry)
}

func (d *ModelDictionary) checkDuplicateLocked(entry ModelEntry) ModelDuplicateCheckResult {
	var similar []ModelEntry

	for _, existing := range d.models {
		if entry.Name == existing.Name {
			e := existing
			e.Similarity = 1.0
			similar = append(similar, e)
			continue
		}

		structuralScore := computeStructuralSimilarity(entry, existing)
		keywordScore := computeKeywordSimilarity(modelKeywords(entry), modelKeywords(existing))
		combinedScore := structuralScore*0.6 + keywordScore*0.4

		if combinedScore >= d.config.SimilarityThreshold {
			e := existing
			e.Similarity = combinedScore
			similar = append(similar, e)
		}
	}

	for i := 0; i < len(similar); i++ {
		for j := i + 1; j < len(similar); j++ {
			if similar[j].Similarity > similar[i].Similarity {
				similar[i], similar[j] = similar[j], similar[i]
			}
		}
	}

	return ModelDuplicateCheckResult{
		IsDuplicate:   len(similar) > 0,
		SimilarModels: similar,
	}
}

// RecordRelationship 记录 Entity-DTO 关系。
func (d *ModelDictionary) RecordRelationship(entity, dto, relType string) {
	d.mu.Lock()
	defer d.mu.Unlock()

	for _, r := range d.relationships {
		if r.Entity == entity && r.DTO == dto && r.Type == relType {
			return
		}
	}
	d.relationships = append(d.relationships, ModelRelationship{
		Entity: entity,
		DTO:    dto,
		Type:   relType,
	})
}

// GetRelationships 获取所有关系。
func (d *ModelDictionary) GetRelationships() []ModelRelationship {
	d.mu.RLock()
	defer d.mu.RUnlock()
	out := make([]ModelRelationship, len(d.relationships))
	copy(out, d.relationships)
	return out
}

// LoadFromYAML 从 YAML 文件加载模型字典。
func (d *ModelDictionary) LoadFromYAML() error {
	d.mu.Lock()
	defer d.mu.Unlock()

	data, err := os.ReadFile(d.config.DictionaryPath)
	if err != nil {
		if os.IsNotExist(err) {
			d.models = make([]ModelEntry, 0)
			d.relationships = make([]ModelRelationship, 0)
			return nil
		}
		return fmt.Errorf("read model dictionary: %w", err)
	}

	var payload struct {
		Models        []ModelEntry        `yaml:"models"`
		Relationships []ModelRelationship `yaml:"relationships"`
	}
	if err := yaml.Unmarshal(data, &payload); err != nil {
		return fmt.Errorf("unmarshal model dictionary: %w", err)
	}

	d.models = payload.Models
	if d.models == nil {
		d.models = make([]ModelEntry, 0)
	}
	d.relationships = payload.Relationships
	if d.relationships == nil {
		d.relationships = make([]ModelRelationship, 0)
	}
	return nil
}

// SaveToYAML 保存模型字典到 YAML 文件。
func (d *ModelDictionary) SaveToYAML() error {
	d.mu.RLock()
	defer d.mu.RUnlock()

	dir := filepath.Dir(d.config.DictionaryPath)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("create dictionary dir: %w", err)
	}

	payload := struct {
		Models        []ModelEntry        `yaml:"models"`
		Relationships []ModelRelationship `yaml:"relationships,omitempty"`
	}{
		Models:        d.models,
		Relationships: d.relationships,
	}

	data, err := yaml.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal model dictionary: %w", err)
	}

	header := []byte("# CodeFlow Model Dictionary\n# Auto-generated - do not edit manually\n\n")
	content := append(header, data...)

	if err := os.WriteFile(d.config.DictionaryPath, content, 0o644); err != nil {
		return fmt.Errorf("write model dictionary: %w", err)
	}
	return nil
}

// GetModels 获取所有模型。
func (d *ModelDictionary) GetModels() []ModelEntry {
	d.mu.RLock()
	defer d.mu.RUnlock()
	out := make([]ModelEntry, len(d.models))
	copy(out, d.models)
	return out
}

// GetModelCount 获取模型数量。
func (d *ModelDictionary) GetModelCount() int {
	d.mu.RLock()
	defer d.mu.RUnlock()
	return len(d.models)
}

func modelKeywords(model ModelEntry) []string {
	parts := []string{model.Name, model.Source}
	parts = append(parts, model.Tags...)
	for _, f := range model.Fields {
		parts = append(parts, f.Name)
	}
	return extractKeywords(strings.Join(parts, " "))
}

func computeStructuralSimilarity(a, b ModelEntry) float64 {
	fieldsA := make(map[string]bool)
	for _, f := range a.Fields {
		fieldsA[f.Name+":"+f.Type] = true
	}

	fieldsB := make(map[string]bool)
	for _, f := range b.Fields {
		fieldsB[f.Name+":"+f.Type] = true
	}

	if len(fieldsA) == 0 && len(fieldsB) == 0 {
		return 0
	}

	intersection := 0
	for k := range fieldsA {
		if fieldsB[k] {
			intersection++
		}
	}

	union := make(map[string]bool)
	for k := range fieldsA {
		union[k] = true
	}
	for k := range fieldsB {
		union[k] = true
	}

	if len(union) == 0 {
		return 0
	}

	jaccardFields := float64(intersection) / float64(len(union))

	// 字段名重叠
	namesA := make(map[string]bool)
	for _, f := range a.Fields {
		namesA[strings.ToLower(f.Name)] = true
	}
	namesB := make(map[string]bool)
	for _, f := range b.Fields {
		namesB[strings.ToLower(f.Name)] = true
	}

	nameIntersection := 0
	for n := range namesA {
		if namesB[n] {
			nameIntersection++
		}
	}

	nameUnion := make(map[string]bool)
	for n := range namesA {
		nameUnion[n] = true
	}
	for n := range namesB {
		nameUnion[n] = true
	}

	nameOverlap := 0.0
	if len(nameUnion) > 0 {
		nameOverlap = float64(nameIntersection) / float64(len(nameUnion))
	}

	return jaccardFields*0.7 + nameOverlap*0.3
}

File: packages/core/src/shadow/ModelDictionary.ts
import * as fs from 'fs';
import * as path from 'path';

export interface ModelField {
  name: string;
  type: string;
  required: boolean;
}

export interface ModelRelationship {
  entity: string;
  dto: string;
  type: 'map' | 'extend' | 'subset' | 'transform';
}

export interface ModelEntry {
  name: string;
  fields: ModelField[];
  relationships: ModelRelationship[];
  source: string;
  tags: string[];
  similarity?: number;
}

export interface ModelDictionaryConfig {
  dictionaryPath: string;
  similarityThreshold: number;
}

interface DuplicateCheckResult {
  isDuplicate: boolean;
  similarModels: ModelEntry[];
}

const DEFAULT_CONFIG: ModelDictionaryConfig = {
  dictionaryPath: '.codeflow/registry/models.yaml',
  similarityThreshold: 0.6,
};

export class ModelDictionary {
  private models: ModelEntry[] = [];
  private relationships: ModelRelationship[] = [];
  private readonly config: ModelDictionaryConfig;

  constructor(config: Partial<ModelDictionaryConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  async register(entry: ModelEntry): Promise<DuplicateCheckResult> {
    const result = this.checkDuplicate(entry);
    if (result.isDuplicate) {
      return result;
    }
    this.models.push({ ...entry, similarity: undefined });
    return { isDuplicate: false, similarModels: [] };
  }

  search(query: string): ModelEntry[] {
    const queryKeywords = this.extractKeywords(query);
    if (queryKeywords.length === 0) return [];

    const scored = this.models
      .map((model) => {
        const modelKw = this.modelKeywords(model);
        const score = this.computeKeywordSimilarity(queryKeywords, modelKw);
        return { model: { ...model, similarity: score }, score };
      })
      .filter((item) => item.score > 0)
      .sort((a, b) => b.score - a.score);

    return scored.map((item) => item.model);
  }

  checkDuplicate(entry: ModelEntry): DuplicateCheckResult {
    const similar: ModelEntry[] = [];

    for (const existing of this.models) {
      if (entry.name === existing.name) {
        similar.push({ ...existing, similarity: 1.0 });
        continue;
      }

      const structuralScore = this.computeStructuralSimilarity(entry, existing);
      const keywordScore = this.computeKeywordSimilarity(
        this.modelKeywords(entry),
        this.modelKeywords(existing),
      );
      const combinedScore = structuralScore * 0.6 + keywordScore * 0.4;

      if (combinedScore >= this.config.similarityThreshold) {
        similar.push({ ...existing, similarity: combinedScore });
      }
    }

    return {
      isDuplicate: similar.length > 0,
      similarModels: similar.sort((a, b) => (b.similarity ?? 0) - (a.similarity ?? 0)),
    };
  }

  recordRelationship(entity: string, dto: string, type: ModelRelationship['type']): void {
    const existing = this.relationships.find(
      (r) => r.entity === entity && r.dto === dto && r.type === type,
    );
    if (!existing) {
      this.relationships.push({ entity, dto, type });
    }
  }

  getRelationships(): ModelRelationship[] {
    return [...this.relationships];
  }

  async loadFromYaml(): Promise<void> {
    const fullPath = path.resolve(this.config.dictionaryPath);
    let content: string;
    try {
      content = await fs.promises.readFile(fullPath, 'utf-8');
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
        this.models = [];
        this.relationships = [];
        return;
      }
      throw err;
    }

    const parsed = this.parseYaml(content);
    this.models = parsed.models;
    this.relationships = parsed.relationships;
  }

  async saveToYaml(): Promise<void> {
    const fullPath = path.resolve(this.config.dictionaryPath);
    await fs.promises.mkdir(path.dirname(fullPath), { recursive: true });
    const content = this.serializeYaml();
    await fs.promises.writeFile(fullPath, content, 'utf-8');
  }

  getModels(): ModelEntry[] {
    return [...this.models];
  }

  getModelCount(): number {
    return this.models.length;
  }

  private computeStructuralSimilarity(a: ModelEntry, b: ModelEntry): number {
    const fieldsA = new Set(a.fields.map((f) => `${f.name}:${f.type}`));
    const fieldsB = new Set(b.fields.map((f) => `${f.name}:${f.type}`));

    if (fieldsA.size === 0 && fieldsB.size === 0) return 0;

    let intersection = 0;
    for (const f of fieldsA) {
      if (fieldsB.has(f)) intersection++;
    }

    const union = new Set([...fieldsA, ...fieldsB]).size;
    if (union === 0) return 0;

    const fieldNameOverlap = this.computeFieldNameOverlap(a.fields, b.fields);

    return (intersection / union) * 0.7 + fieldNameOverlap * 0.3;
  }

  private computeFieldNameOverlap(fieldsA: ModelField[], fieldsB: ModelField[]): number {
    const namesA = new Set(fieldsA.map((f) => f.name.toLowerCase()));
    const namesB = new Set(fieldsB.map((f) => f.name.toLowerCase()));

    if (namesA.size === 0 && namesB.size === 0) return 0;

    let intersection = 0;
    for (const n of namesA) {
      if (namesB.has(n)) intersection++;
    }

    const union = new Set([...namesA, ...namesB]).size;
    return union === 0 ? 0 : intersection / union;
  }

  private modelKeywords(model: ModelEntry): string[] {
    const parts = [model.name, model.source, ...model.tags, ...model.fields.map((f) => f.name)];
    return this.extractKeywords(parts.join(' '));
  }

  private extractKeywords(text: string): string[] {
    return text
      .toLowerCase()
      .replace(/[^\w\s\u4e00-\u9fff]/g, ' ')
      .split(/\s+/)
      .filter((t) => t.length > 1);
  }

  private computeKeywordSimilarity(keywordsA: string[], keywordsB: string[]): number {
    if (keywordsA.length === 0 || keywordsB.length === 0) return 0;

    const setA = new Set(keywordsA);
    const setB = new Set(keywordsB);

    let intersection = 0;
    for (const kw of setA) {
      if (setB.has(kw)) intersection++;
    }

    const union = new Set([...setA, ...setB]).size;
    return union === 0 ? 0 : intersection / union;
  }

  private parseYaml(content: string): { models: ModelEntry[]; relationships: ModelRelationship[] } {
    const models: ModelEntry[] = [];
    const relationships: ModelRelationship[] = [];

    const sections = content.split(/^# /m);

    for (const section of sections) {
      const trimmed = section.trim();
      if (trimmed.startsWith('Models') || trimmed.startsWith('models')) {
        const blocks = trimmed.split(/^- name:/m).filter((b) => b.trim().length > 0);
        for (const block of blocks) {
          if (!block.includes(':')) continue;
          const model = this.parseModelBlock(`name:${block}`);
          if (model) models.push(model);
        }
      } else if (trimmed.startsWith('Relationships') || trimmed.startsWith('relationships')) {
        const relBlocks = trimmed.split(/^- /m).filter((b) => b.includes('entity:'));
        for (const block of relBlocks) {
          const rel = this.parseRelationshipBlock(block);
          if (rel) relationships.push(rel);
        }
      }
    }

    return { models, relationships };
  }

  private parseModelBlock(block: string): ModelEntry | null {
    const lines = block.split('\n');
    let name = '';
    let source = '';
    const tags: string[] = [];
    const fields: ModelField[] = [];

    for (const line of lines) {
      const trimmed = line.trim();
      if (trimmed.startsWith('name:')) name = trimmed.slice(5).trim().replace(/^["']|["']$/g, '');
      else if (trimmed.startsWith('source:')) source = trimmed.slice(7).trim().replace(/^["']|["']$/g, '');
      else if (trimmed.startsWith('tags:')) {
        const tagsStr = trimmed.slice(5).trim();
        if (tagsStr.startsWith('[')) {
          tags.push(
            ...tagsStr
              .replace(/^\[|\]$/g, '')
              .split(',')
              .map((t) => t.trim().replace(/^["']|["']$/g, ''))
              .filter(Boolean),
          );
        }
      } else if (trimmed.startsWith('- field:')) {
        const fieldParts = trimmed.slice(8).trim().split(',');
        if (fieldParts.length >= 2) {
          fields.push({
            name: fieldParts[0].trim(),
            type: fieldParts[1].trim(),
            required: fieldParts[2]?.trim() === 'required',
          });
        }
      }
    }

    if (!name) return null;
    return { name, fields, relationships: [], source, tags };
  }

  private parseRelationshipBlock(block: string): ModelRelationship | null {
    let entity = '';
    let dto = '';
    let type: ModelRelationship['type'] = 'map';

    for (const line of block.split('\n')) {
      const trimmed = line.trim();
      if (trimmed.startsWith('entity:')) entity = trimmed.slice(7).trim().replace(/^["']|["']$/g, '');
      else if (trimmed.startsWith('dto:')) dto = trimmed.slice(4).trim().replace(/^["']|["']$/g, '');
      else if (trimmed.startsWith('type:')) {
        const val = trimmed.slice(5).trim().replace(/^["']|["']$/g, '') as ModelRelationship['type'];
        if (['map', 'extend', 'subset', 'transform'].includes(val)) type = val;
      }
    }

    if (!entity || !dto) return null;
    return { entity, dto, type };
  }

  private serializeYaml(): string {
    const lines: string[] = [
      '# CodeFlow Model Dictionary',
      '# Auto-generated - do not edit manually',
      '',
      '# Models',
    ];

    for (const model of this.models) {
      lines.push(`- name: "${model.name}"`);
      lines.push(`  source: "${model.source}"`);
      lines.push(`  tags: [${model.tags.map((t) => `"${t}"`).join(', ')}]`);
      if (model.fields.length > 0) {
        lines.push('  fields:');
        for (const field of model.fields) {
          lines.push(`    - field: ${field.name}, ${field.type}, ${field.required ? 'required' : 'optional'}`);
        }
      }
      lines.push('');
    }

    if (this.relationships.length > 0) {
      lines.push('# Relationships');
      for (const rel of this.relationships) {
        lines.push(`- entity: "${rel.entity}"`);
        lines.push(`  dto: "${rel.dto}"`);
        lines.push(`  type: "${rel.type}"`);
        lines.push('');
      }
    }

    return lines.join('\n');
  }
}
